honeybadgermpc/avss_value_processor.py

```diff
@@ -1,85 +1,104 @@
+            outputs = self.outputs_per_dealer[dealer_id]
 
-    async def _recv_loop(self):
-        logging.debug("[%d] Starting recv_loop", self.my_id)
+            # If we have the Future available for the next value to be returned
+            # then return a reference to that future and increment the next idx
+            # otherwise keep waiting.
+            if next_idx < len(outputs):
+                logging.debug("[%d] Future available, dealer_id: %d, next idx: %d",
+                              self.my_id, dealer_id, next_idx)
+                self.next_idx_to_return_per_dealer[dealer_id] += 1
+                return self.outputs_per_dealer[dealer_id][next_idx]
 
+            await asyncio.sleep(AvssValueProcessor.WAIT_PERIOD_IN_SECONDS)
+
+    async def _recv_loop(self):
+        logging.debug("[%d] Starting _recv_loop", self.my_id)
         while True:
-            dealer_id, avss_value = await self.get_input()
+            dealer_id, avss_id, avss_value = await self.get_input()
             assert type(dealer_id) is int
             assert dealer_id >= 0 and dealer_id < self.n
+            assert type(avss_id) is int
+            assert avss_id >= 0
+
+            self.sequencers[dealer_id].add((avss_id, avss_value))
+
+            # Process the values only if they are in order.
+            while self.sequencers[dealer_id].is_next_available():
+                _, avss_value = self.sequencers[dealer_id].get()
 
-            # Add the value to the input list based on who dealt the value
-            self.inputs_per_dealer[dealer_id].append(avss_value)
+                # Add the value to the input list based on who dealt the value
+                self.inputs_per_dealer[dealer_id].append(avss_value)
 
-            # If this value has already been agreed upon by other parties
-            # then it means that its Future has been added to the output list.
-            # So we need to set the result of that future to be equal to this value.
-            idx = len(self.inputs_per_dealer[dealer_id])-1
-            if idx < len(self.outputs_per_dealer[dealer_id]):
-                assert not self.outputs_per_dealer[dealer_id][idx].done()
-                self.outputs_per_dealer[dealer_id][idx].set_result(avss_value)
+                # If this value has already been agreed upon by other parties
+                # then it means that its Future has been added to the output list.
+                # So we need to set the result of that future to be equal to this value.
+                idx = len(self.inputs_per_dealer[dealer_id])-1
+                if idx < len(self.outputs_per_dealer[dealer_id]):
+                    assert not self.outputs_per_dealer[dealer_id][idx].done()
+                    self.outputs_per_dealer[dealer_id][idx].set_result(avss_value)
+
+    async def _acs_runner(self):
+        logging.debug("[%d] Starting ACS runner", self.my_id)
+        acs_counter = 0
+        while True:
```

Comment:
I'd like to iterate a bit on the policy for how we decide when to start the ACS next. Having a timeout is great. But another policy can be to wait until some new completed values have been added to the AVSS queues. I think there is not a perfect answer here, just heuristics, but we can discuss some options

Reply:
Sure, we can definitely be smarter about how we do this. In fact, we can have both - a timeout along with doing an ACS only when we have some available values or more than `k` values.

Opened #218 to track this.